s3/src/test/java/ch/cyberduck/core/sts/OidcAuthenticationTest.java

```diff
@@ -0,0 +1,168 @@
+        session.login(Proxy.DIRECT, new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.close();
+    }
+
+    @Test
+    public void testTokenRefresh() throws BackgroundException, InterruptedException {
+        final Host host = new Host(profile, profile.getDefaultHostname(), new Credentials("rawuser", "rawuser"));
+        final S3Session session = new S3Session(host);
+        session.open(Proxy.DIRECT, new DisabledHostKeyCallback(), new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.login(Proxy.DIRECT, new DisabledLoginCallback(), new DisabledCancelCallback());
+
+        String firstAccessToken = session.getAuthorizationService().getTokens().getIdToken();
+        String firstAccessKey = session.getClient().getProviderCredentials().getAccessKey();
+        String firstSessionToken = ((AWSSessionCredentials) session.getClient().getProviderCredentials()).getSessionToken();
+
+        Path container = new Path("cyberduckbucket", EnumSet.of(Path.Type.directory, Path.Type.volume));
+        assertTrue(new S3FindFeature(session, new S3AccessControlListFeature(session)).find(container));
+        Thread.sleep(1100 * 30);
+        assertTrue(new S3FindFeature(session, new S3AccessControlListFeature(session)).find(container));
+
+        assertNotEquals(firstAccessToken, session.getAuthorizationService().getTokens().getIdToken());
+        assertNotEquals(firstAccessKey, session.getClient().getProviderCredentials().getAccessKey());
+        assertNotEquals(firstSessionToken, ((AWSSessionCredentials) session.getClient().getProviderCredentials()).getSessionToken());
+        session.close();
+    }
+
+    @Test
+    public void testSTSCredentialExpiryTimeIsBoundToOAuthExpiryTime() throws BackgroundException {
+        final Host host = new Host(profile, profile.getDefaultHostname(), new Credentials("rawuser", "rawuser"));
+        host.setProperty("s3.assumerole.durationseconds", "900");
+        assertEquals(new HostPreferences(host).getInteger("s3.assumerole.durationseconds"), 900);
+        final S3Session session = new S3Session(host);
+        session.open(Proxy.DIRECT, new DisabledHostKeyCallback(), new DisabledLoginCallback(), new DisabledCancelCallback());
+        session.login(Proxy.DIRECT, new DisabledLoginCallback(), new DisabledCancelCallback());
+
+        // assert that STS credentials expires with the oAuth token even though the duration seconds is valid for longer
+        STSCredentialsRequestInterceptor authorizationService = session.getAuthorizationService();
+        assertTrue(40 > ((authorizationService.getStsExpiryInMilliseconds() - System.currentTimeMillis()) / 1000));
+        assertEquals(Optional.of(authorizationService.getStsExpiryInMilliseconds()).get() / 1000, authorizationService.getTokens().getExpiryInMilliseconds() / 1000);
+
+        session.close();
+    }
+
+    /** only use with the below specified changes in the keycloak config json file and run as separate test
```

Comment:
Ideally we find a way to run these tests with different configurations in `keycloak-realm.json`.

Reply:
My first thought was to use a second realm with these configurations, but that would lead to a second IDP config in MinIO which is not possible with the same address. As of now I do not know how to do this without maybe using additional containers, which would complicate the Testcontainers and the Docker config.

Reply:
@dkocher I have drafted a way to run test with different configurations in `keycloak-realm.json`.
`testSTSCredentialsExpiredValidOAuthToken` will take 15 minutes as the min STS duration is 900s see https://min.io/docs/minio/linux/developers/security-token-service/AssumeRoleWithWebIdentity.html. So unsure what it makes sense to finish it.

@IsmCadi I struggle to see the  working of `testBucketRequestBeforeTokenExpiryFailsBecauseOfLatency` - can you re-phrase what you're testing and how to adjust the 28 secs? The test is red when I run locally.

Reply:
Let's continue discussion here: https://github.com/iterate-ch/cyberduck/pull/14972/files#r1291199162